### PR TITLE
doc: add section on breaking changes to contribution guidelines

### PR DESCRIPTION
**Pull request by yaauie**
Attempt to document the observed _ethos_ with regard to avoiding and embracing breaking changes.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -63,6 +63,24 @@ https://github.com/elastic/logstash/labels/low%20hanging%20fruit
...
+
+When we do so, we need to acknowledge the work we are placing on the rest of our users the next time they upgrade, and work to ensure that they can upgrade confidently.
+
+Where possible, we:
+ 1. first implement new behaviour in a way that users can explicitly opt into (MINOR),
+ 2. introduce deprecation warnings when old behaviour is used, communicating the pending change (MINOR, potentially along-side #1),
+ 3. change the default to be new behaviour, optionally allowing users to opt out in favor of the old behaviour (MAJOR), and eventually
+ 4. remove the old behaviour's implementation from the code-base (MAJOR, potentially along-side #3).
+
+If you request or submit a change that is flagged as a "breaking change", we have several paths forward:
```
Comment (yaauie) [inline CONTRIBUTING.md]: not confident on the voice of this line, especially how it aligns with the rest of the document. I want to call out that a breaking change isn't the end of the road (without using idioms like "end of the road"), and that contributions with breaking changes aren't just flat-out being rejected.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -63,6 +63,24 @@ https://github.com/elastic/logstash/labels/low%20hanging%20fruit
...
+
+When introducing new behaviour, we favor implementing it in a non-breaking way that users can opt into, meaning users' existing configurations continue to work as they expect them to after upgrading Logstash.
+
+But sometimes it is necessary to introduce "breaking changes", whether that is removing an option that doesn't make sense anymore, changing the default value of a setting, or reimplementing a major component differently for performance or safety reasons.
+
+When we do so, we need to acknowledge the work we are placing on the rest of our users the next time they upgrade, and work to ensure that they can upgrade confidently.
+
+Where possible, we:
+ 1. first implement new behaviour in a way that users can explicitly opt into (MINOR),
+ 2. introduce deprecation warnings when old behaviour is used, communicating the pending change (MINOR, potentially along-side #1),
```
Comment (yaauie) [inline CONTRIBUTING.md]: deprecation warnings are only one way to communicate the pending change; maybe reword to make the _communication_ be the primary goal.

```suggestion
 2. commuicate the pending change in behaviour, including the introduction of deprecation warnings when old behaviour is used (MINOR, potentially along-side #1),
```

File: CONTRIBUTING.md (hunk 1 above)
Comment (jsvd) [inline CONTRIBUTING.md]: Maybe:
```suggestion
Once a pull request is marked as a "breaking change", it becomes necessary to either:
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -63,6 +63,24 @@ https://github.com/elastic/logstash/labels/low%20hanging%20fruit
...
+ 3. change the default to be new behaviour, optionally allowing users to opt out in favor of the old behaviour (MAJOR), and eventually
```
Comment (jsvd) [inline CONTRIBUTING.md]: at this point we should also note that a entry should be added to the next major's breaking change document https://github.com/elastic/logstash/blob/master/docs/static/breaking-changes.asciidoc, what do you think?

Comment (yaauie) [reply CONTRIBUTING.md]:
```suggestion
 3. change the default to be new behaviour, communicating the breaking change and optionally allowing users to opt out in favor of the old behaviour (MAJOR), and eventually
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -63,6 +63,24 @@ https://github.com/elastic/logstash/labels/low%20hanging%20fruit
 
 Using IntelliJ? See a detailed getting started guide [here](https://docs.google.com/document/d/1kqunARvYMrlfTEOgMpYHig0U-ZqCcMJfhvTtGt09iZg/pub).
 
+## Breaking Changes
+
+When introducing new behaviour, we favor implementing it in a non-breaking way that users can opt into, meaning users' existing configurations continue to work as they expect them to after upgrading Logstash.
+
+But sometimes it is necessary to introduce "breaking changes", whether that is removing an option that doesn't make sense anymore, changing the default value of a setting, or reimplementing a major component differently for performance or safety reasons.
```
Comment (karenzone) [inline CONTRIBUTING.md]:
```suggestion
But sometimes it is necessary to introduce "breaking changes," whether that is removing an option that doesn't make sense anymore, changing the default value of a setting, or reimplementing a major component differently for performance or safety reasons.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -63,6 +63,24 @@ https://github.com/elastic/logstash/labels/low%20hanging%20fruit
...
+
+When we do so, we need to acknowledge the work we are placing on the rest of our users the next time they upgrade, and work to ensure that they can upgrade confidently.
+
+Where possible, we:
+ 1. first implement new behaviour in a way that users can explicitly opt into (MINOR),
+ 2. introduce deprecation warnings when old behaviour is used, communicating the pending change (MINOR, potentially along-side #1),
+ 3. change the default to be new behaviour, optionally allowing users to opt out in favor of the old behaviour (MAJOR), and eventually
+ 4. remove the old behaviour's implementation from the code-base (MAJOR, potentially along-side #3).
+
+Once a pull request is marked as a "breaking change", it becomes necessary to either:
```
Comment (karenzone) [inline CONTRIBUTING.md]:
```suggestion
After a pull request is marked as a "breaking change," it becomes necessary to either:
```

File: CONTRIBUTING.md (hunk 3 above)
Comment (karenzone) [reply CONTRIBUTING.md]: Consider  keeping verbs parallel "change," "communicate," and "allow"

Comment (yaauie) [reply CONTRIBUTING.md:77]:
```suggestion
 3. change the default to be new behaviour, communicate the breaking change, optionally allow users to opt out in favor of the old behaviour (MAJOR), and eventually
```

**Review by karenzone — APPROVED** [review]
Thank you for moving this forward. LGTM
